### PR TITLE
core/rawdb: find smallest block stored in key-value store when chain gapped

### PR DESCRIPTION
**Pull request by rjl493456442**

### CODE REVIEW

**Comment by holiman** [comment]
I added more printouts, in case geth fails to start up due to "something strange in the database". 


When I intentionally cause an error (mismatched ancient and leveldb):
```
[user@work go-ethereum]$ go run ./cmd/geth/ --datadir.ancient=/home/user/.ethereum/ropsten/geth/chaindata/ancient/
INFO [02-17|10:02:55.755] Starting Geth on Ethereum mainnet... 
INFO [02-17|10:02:55.755] Bumping default cache on mainnet         provided=1024 updated=4096
INFO [02-17|10:02:55.757] Maximum peer count                       ETH=50 LES=0 total=50
INFO [02-17|10:02:55.783] Set global gas cap                       cap=50,000,000
INFO [02-17|10:02:55.786] Allocated trie memory caches             clean=614.00MiB dirty=1024.00MiB
INFO [02-17|10:02:55.786] Using leveldb as the backing database 
INFO [02-17|10:02:55.786] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=2.00GiB handles=262,144
INFO [02-17|10:02:55.952] Using LevelDB as the backing database 
INFO [02-17|10:02:55.952] Found legacy ancient chain path          location=/home/user/.ethereum/ropsten/geth/chaindata/ancient
INFO [02-17|10:02:55.962] Opened ancient database                  database=/home/user/.ethereum/ropsten/geth/chaindata/ancient readonly=false
Chain metadata
  databaseVersion : 8 (0x8)
  headBlockHash : 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3
  headFastBlockHash : 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3
  headHeaderHash : 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3
  lastPivotNumber : <nil>
  len(snapshotSyncStatus) : 0 bytes
  snapshotDisabled : false
  snapshotJournal : 34 bytes
  snapshotRecoveryNumber : <nil>
  snapshotRoot : 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544
  txIndexTail : <nil>
  fastTxLookupLimit : <nil>


Fatal: Failed to register the Ethereum service: genesis mismatch: 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 (leveldb) != 0x41941023680923e0fe4d74a34bdac8141f2540e3ae90623718e47d66d1ca4a2d (ancients)

```
Verification that I didn't break `db metadata`: 
```
[user@work go-ethereum]$ go run ./cmd/geth db metadata
INFO [02-17|10:04:41.835] Maximum peer count                       ETH=50 LES=0 total=50
INFO [02-17|10:04:41.856] Set global gas cap                       cap=50,000,000
INFO [02-17|10:04:41.857] Using leveldb as the backing database 
INFO [02-17|10:04:41.857] Allocated cache and file handles         database=/home/user/.ethereum/geth/chaindata cache=512.00MiB handles=262,144 readonly=true
INFO [02-17|10:04:41.864] Using LevelDB as the backing database 
INFO [02-17|10:04:41.864] Found legacy ancient chain path          location=/home/user/.ethereum/geth/chaindata/ancient
INFO [02-17|10:04:41.879] Opened ancient database                  database=/home/user/.ethereum/geth/chaindata/ancient readonly=true
+-------------------------+--------------------------------------------------------------------+
|          FIELD          |                               VALUE                                |
+-------------------------+--------------------------------------------------------------------+
| databaseVersion         | 8 (0x8)                                                            |
| headBlockHash           | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headFastBlockHash       | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headHeaderHash          | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| lastPivotNumber         | <nil>                                                              |
| len(snapshotSyncStatus) | 0 bytes                                                            |
| snapshotDisabled        | false                                                              |
| snapshotJournal         | 34 bytes                                                           |
| snapshotRecoveryNumber  | <nil>                                                              |
| snapshotRoot            | 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544 |
| txIndexTail             | <nil>                                                              |
| fastTxLookupLimit       | <nil>                                                              |
| frozen                  | 0 items                                                            |
| snapshotGenerator       | Done: true, Accounts: 0,                                           |
|                         | Slots: 0, Storage: 0, Marker:                                      |
| headBlock.Hash          | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headBlock.Root          | 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544 |
| headBlock.Number        | 0 (0x0)                                                            |
| headHeader.Hash         | 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3 |
| headHeader.Root         | 0xd7f8974fb5ac78d9ac099b9ad5018bedc2ce0a72dad1827a1709da30580f0544 |
| headHeader.Number       | 0 (0x0)                                                            |
+-------------------------+--------------------------------------------------------------------+
```

**Comment by holiman** [comment]
Made the check ascending now, and the printout a bit clearer (I hope). Need to check that I haven't off-by-one:d something

File: core/rawdb/database.go [hunk 1]
```diff
@@ -581,3 +596,42 @@ func InspectDatabase(db ethdb.Database, keyPrefix, keyStart []byte) error {
 	}
 	return nil
 }
+
+// PrintChainMetadata prints out chain metadata to stderr
+func PrintChainMetadata(db ethdb.KeyValueStore) {
```
Comment (holiman) [inline core/rawdb/database.go:601]: unexport

**Review by holiman — APPROVED** [review]
LGTM, but I wrote parts of it. @rjl493456442 lgty?

**Review by rjl493456442 — COMMENTED** [review]
LGTM

**Comment by BrittneyNi** [comment]
Why are you hooked to my email, and who are all you? One you tough guys
hacked a single mom hope ya all feel super tough about yourselves that your
only where your at in life because you robbed and stole peoples money and
identity you will never shine when you don’t build something on your
own.Goodluck!

**Pull request merged**